File: src/io.c
#include <kernel.h>

/*
 * Inputs a byte from the specified port.
 */
ATTR_HOT
uint8_t inb(const uint16_t port)
{
	uint8_t ret;
	asm volatile("inb %1, %0" : "=a"(ret) : "d"(port));

	return ret;
}

/*
 * Inputs a word from the specified port.
 */
ATTR_HOT
uint16_t inw(const uint16_t port)
{
	uint16_t ret;
	asm volatile("inw %1, %0" : "=a"(ret) : "d"(port));

	return ret;
}

/*
 * Inputs a long from the specified port.
 */
ATTR_HOT
uint32_t inl(uint16_t port)
{
	uint32_t ret;
	asm volatile("inl %1, %0" : "=a"(ret) : "d"(port));

	return ret;
}

/*
 * Outputs a byte to the specified port.
 */
ATTR_HOT
void outb(const uint16_t port, const uint8_t value)
{
	asm volatile("outb %0, %1" : : "a"(value), "d"(port));
}

/*
 * Outputs a word to the specified port.
 */
ATTR_HOT
void outw(const uint16_t port, const uint16_t value)
{
	asm volatile("outw %0, %1" : : "a"(value), "d"(port));
}

/*
 * Outputs a long to the specified port.
 */
ATTR_HOT
void outl(const uint16_t port, const uint32_t value)
{
	asm volatile("outl %0, %1" : : "a"(value), "d"(port));
}

File: src/io.rs
//! The I/O functions allow to communicate with the other components on the system.

use core::arch::asm;

/// Inputs a byte from the specified port.
pub unsafe fn inb(port: u16) -> u8 {
	let ret: i8;
	asm!("in al, dx", out("al") ret, in("dx") port);

	ret as _
}

/// Inputs a word from the specified port.
pub unsafe fn inw(port: u16) -> u16 {
	let ret: i16;
	asm!("in ax, dx", out("ax") ret, in("dx") port);

	ret as _
}

/// Inputs a long from the specified port.
pub unsafe fn inl(port: u16) -> u32 {
	let ret: i32;
	asm!("in eax, dx", out("eax") ret, in("dx") port);

	ret as _
}

/// Outputs a byte to the specified port.
pub unsafe fn outb(port: u16, value: u8) {
	asm!("out dx, al", in("al") value, in("dx") port);
}

/// Outputs a word to the specified port.
pub unsafe fn outw(port: u16, value: u16) {
	asm!("out dx, ax", in("ax") value, in("dx") port);
}

/// Outputs a long to the specified port.
pub unsafe fn outl(port: u16, value: u32) {
	asm!("out dx, eax", in("eax") value, in("dx") port);
}
